### PR TITLE
[BEAM-2914] Add TimestampPrefixingWindowCoder as beam:coder:custom_window:v1.

### PR DESCRIPTION
**Pull request by y1chi**
Add a new common standard coder beam:coder:custom_window:v1. The coder encodes arbitrary bounded window by prefixing the max timestamp to its encoded form using the window's registered coder. e.g.:
   max_timestamp
   encoded_window_bytes

The idea is that, any arbitrary window extending bounded window must have max timestamp defined and runner only needs to know the max timestamp when processing trigger timers. The encoded bytes using the original custom window's registered coder can be used transparently on runner side as the identity of the window.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by y1chi** [comment]
Run Java PreCommit

**Comment by y1chi** [comment]
Run GoPortable PreCommit

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/CommonCoderTest.java [hunk 1]
```diff
@@ -345,6 +347,11 @@ private static Object convertValue(Object value, CommonCoder coderSpec, Coder co
       byte[] shardId = ((String) kvMap.get("shardId")).getBytes(StandardCharsets.ISO_8859_1);
       return ShardedKey.of(
           convertValue(kvMap.get("key"), coderSpec.getComponents().get(0), keyCoder), shardId);
+    } else if (s.equals(getUrn(StandardCoders.Enum.CUSTOM_WINDOW))) {
+      Map<String, Object> kvMap = (Map<String, Object>) value;
+      Instant end = new Instant(((Number) kvMap.get("end")).longValue());
+      Duration span = Duration.millis(((Number) kvMap.get("span")).longValue());
+      return new IntervalWindow(end.minus(span), span);
```
Comment (lukecwik) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/CommonCoderTest.java:354]: Other places we allow any arbitrary type to be placed for the component, so how about we update this to something like:
```
Coder windowCoder = ((TimestampPrefixingWindowCoder) coder).getWindowCoder();
Object k = convertValue(kvMap.get("window"), coderSpec.getComponents().get(0), windowCoder);
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoder.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    return windowCoder.decode(inStream);
+  }
+
+  @Override
+  public List<? extends Coder<?>> getCoderArguments() {
+    return Lists.newArrayList(windowCoder);
+  }
+
+  @Override
+  public void verifyDeterministic() throws NonDeterministicException {
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoder.java:67]: Please also implement:
* consistentWithEquals (e.g. true iff delegate window coder is consistent with equals)
* isRegisterByteSizeObserverCheap
* registerByteSizeObserver

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    private static final VarIntCoder VAR_INT_CODER = VarIntCoder.of();
+
+    public static CustomWindowCoder of() {
+      return INSTANCE;
+    }
+
+    @Override
+    public void encode(CustomWindow window, OutputStream outStream) throws IOException {
+      INTERVAL_WINDOW_CODER.encode(window, outStream);
+      VAR_INT_CODER.encode(window.isBig ? 1 : 0, outStream);
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java:75]: nit: I think you can rely on BooleanCoder.of() instead

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        Lists.newArrayList(
+            new CustomWindow(new Instant(0L), new Instant(1L), true),
+            new CustomWindow(new Instant(100L), new Instant(200L), false),
+            new CustomWindow(new Instant(0L), BoundedWindow.TIMESTAMP_MAX_VALUE, true));
+    TimestampPrefixingWindowCoder<CustomWindow> coder3 =
+        TimestampPrefixingWindowCoder.of(CustomWindowCoder.of());
+    for (CustomWindow window : customWindowsToTest) {
+      CoderProperties.coderDecodeEncodeEqual(coder3, window);
+    }
+  }
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java:120]: Please add tests for the consistent with equals and cheap observer stuff

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoder.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Lists;
+
+/**
+ * A {@link TimestampPrefixingWindowCoder} wraps arbitrary user custom window coder. While encoding
+ * the custom window type, it extracts the maxTimestamp(inclusive) of the window and prefix it to
+ * the encoded bytes of the window using the user custom window coder.
+ *
+ * @param <T> The custom window type.
+ */
+public class TimestampPrefixingWindowCoder<T> extends StructuredCoder<T> {
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoder.java:34]: T extend BoundedWindow?

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 6]
```diff
@@ -939,6 +939,15 @@ message StandardCoders {
...
     STATE_BACKED_ITERABLE = 9 [(beam_urn) = "beam:coder:state_backed_iterable:v1"];
 
+
+    // Encodes an arbitrary user defined window, the coder extracts the required
+    // maxTimestamp(inclusive) parameter for the window and prefix it to the
+    // bytes encoded from the window by using user defined and registered window
+    // coder.
+    //
+    // Components: Coder for the custom window type.
+    CUSTOM_WINDOW = 16 [(beam_urn) = "beam:coder:custom_window:v1"];
```
[the remark this answers is not in the document]
Comment (y1chi) [reply model/pipeline/src/main/proto/beam_runner_api.proto]: Done.

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/CommonCoderTest.java (hunk 1 above)
Comment (y1chi) [reply runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/CommonCoderTest.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoder.java (hunk 2 above)
Comment (y1chi) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoder.java:74]: Done.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java (hunk 3 above)
Comment (y1chi) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java]: Done.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java (hunk 4 above)
Comment (y1chi) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java:170]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoder.java (hunk 5 above)
Comment (y1chi) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoder.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoder.java [hunk 7]
```diff
@@ -67,4 +74,20 @@ public T decode(InputStream inStream) throws CoderException, IOException {
...
+  }
+
+  @Override
+  public boolean isRegisterByteSizeObserverCheap(T value) {
+    return windowCoder.isRegisterByteSizeObserverCheap(value);
+  }
+
+  @Override
+  public void registerByteSizeObserver(T value, ElementByteSizeObserver observer) throws Exception {
+    InstantCoder.of().registerByteSizeObserver(new Instant(), observer);
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoder.java:90]: Did you mean to pass in the max timestamp of value instead of `new Instant()`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/IntervalWindow.java [hunk 8]
```diff
@@ -174,6 +175,19 @@ public boolean consistentWithEquals() {
       return instantCoder.consistentWithEquals() && durationCoder.consistentWithEquals();
     }
 
+    @Override
+    public boolean isRegisterByteSizeObserverCheap(IntervalWindow value) {
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/IntervalWindow.java:179]: Great!

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java [hunk 9]
```diff
@@ -107,15 +140,32 @@ public void testEncodeAndDecode() throws Exception {
...
+
+  @Test
+  public void testConsistentWithEquals() {
+    assertThat(TEST_CODER.consistentWithEquals(), equalTo(true));
+  }
+
+  @Test
+  public void testIsRegisterByteSizeObserverCheap() {
+    assertThat(
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java:155]: You'll also want to add a variant showing that if the window coder returns false then this is false.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java [hunk 10]
```diff
@@ -107,15 +140,32 @@ public void testEncodeAndDecode() throws Exception {
...
-    for (CustomWindow window : customWindowsToTest) {
-      CoderProperties.coderDecodeEncodeEqual(coder3, window);
+    for (CustomWindow window : CUSTOM_WINDOW_LIST) {
+      CoderProperties.coderDecodeEncodeEqual(TEST_CODER, window);
+    }
+  }
+
+  @Test
+  public void testConsistentWithEquals() {
+    assertThat(TEST_CODER.consistentWithEquals(), equalTo(true));
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java:150]: You'll also want to add a variant showing that if the window coder returns false then this is false.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java [hunk 11]
```diff
@@ -107,15 +140,32 @@ public void testEncodeAndDecode() throws Exception {
...
+  }
+
+  @Test
+  public void testGetEncodedElementByteSize() throws Exception {
+    TestElementByteSizeObserver observer = new TestElementByteSizeObserver();
+    for (CustomWindow value : CUSTOM_WINDOW_LIST) {
+      TEST_CODER.registerByteSizeObserver(value, observer);
+      observer.advance();
+      assertThat(
+          observer.getSumAndReset(),
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java:166]: To verify that the CustomWindowCoder is being used, you should have it "lie" about the size so you know that it isn't the true encoded size which is the default implementation of registerByteSizeObserver.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/IntervalWindow.java (hunk 8 above)
Comment (lukecwik) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/IntervalWindow.java:179]: Add tests to prevent regression?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoder.java (hunk 7 above)
Comment (y1chi) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoder.java]: done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/IntervalWindow.java (hunk 8 above)
Comment (y1chi) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/IntervalWindow.java:179]: done.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java (hunk 9 above)
Comment (y1chi) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java]: done.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java (hunk 11 above)
Comment (y1chi) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java:179]: done.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java (hunk 10 above)
Comment (y1chi) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java]: done.

**Review by lukecwik — APPROVED** [review]

**Comment by lukecwik** [comment]
Run Java PreCommit

**Comment by y1chi** [comment]
Run Java PreCommit

File: model/fn-execution/src/main/resources/org/apache/beam/model/fnexecution/v1/standard_coders.yaml [hunk 12]
```diff
@@ -440,3 +441,25 @@ examples:
...
+#  IntervalWindow window = new IntervalWindow(end.minus(span), span);
+#  byte[] bytes = CoderUtils.encodeToByteArray(coder, window);
+#  String str = new String(bytes, java.nio.charset.StandardCharsets.ISO_8859_1);
+#  String example = "";
+#  for(int i = 0; i < str.length(); i++){
+#    example += CharUtils.unicodeEscaped(str.charAt(i));
+#  }
+#  System.out.println(example);
+coder:
+  urn: "beam:coder:custom_window:v1"
```
Comment (robertwb) [inline model/fn-execution/src/main/resources/org/apache/beam/model/fnexecution/v1/standard_coders.yaml:460]: Should this be timestamp_prefixed (similar to length_prefixed)?

File: model/fn-execution/src/main/resources/org/apache/beam/model/fnexecution/v1/standard_coders.yaml [hunk 13]
```diff
@@ -440,3 +441,25 @@ examples:
...
+  components: [{urn: "beam:coder:interval_window:v1"}]
```
Comment (robertwb) [inline model/fn-execution/src/main/resources/org/apache/beam/model/fnexecution/v1/standard_coders.yaml:461]: This shouldn't be limited to interval_windows, any windowed_coder should do.

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 14]
```diff
@@ -939,6 +939,31 @@ message StandardCoders {
...
 
+
+    // Encodes an arbitrary user defined window and its max timestamp (inclusive).
+    // The encoding format is:
+    //   maxTimestamp window
+    //
+    //   maxTimestamp - A big endian 8 byte integer representing millis-since-epoch.
+    //     The encoded representation is shifted so that the byte representation
+    //     of negative values are lexicographically ordered before the byte
+    //     representation of positive values. This is typically done by
```
Comment (robertwb) [inline model/pipeline/src/main/proto/beam_runner_api.proto:950]: "typically done" makes it sound like any transformation preserving lexicographic ordering is acceptable. There is no choice here. 

Perhaps we could reference other coders here (e.g. the encoding used for TimestmpCoder, or in WindowedValueCoder?

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 15]
```diff
@@ -939,6 +939,31 @@ message StandardCoders {
     // Experimental.
     STATE_BACKED_ITERABLE = 9 [(beam_urn) = "beam:coder:state_backed_iterable:v1"];
 
+
+    // Encodes an arbitrary user defined window and its max timestamp (inclusive).
+    // The encoding format is:
+    //   maxTimestamp window
```
Comment (robertwb) [inline model/pipeline/src/main/proto/beam_runner_api.proto:945]: How is the window encoded? Should we length prefix it as well? Or do we expect TimestampPrefixed(LenghtPrefixed(CustomWindowCoder))?

File: sdks/python/apache_beam/coders/coders.py [hunk 16]
```diff
@@ -1509,3 +1509,47 @@ def __repr__(self):
...
+  def to_type_hint(self):
+    return self._window_coder.to_type_hint()
+
+  def _get_component_coders(self) -> List[Coder]:
+    return [self._window_coder]
+
+  def is_deterministic(self) -> bool:
+    return self._window_coder.is_deterministic()
+
+  def as_cloud_object(self, coders_context=None):
```
Comment (robertwb) [inline sdks/python/apache_beam/coders/coders.py:1535]: Is this needed?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectKinds.java [hunk 17]
```diff
@@ -21,6 +21,7 @@
 class CloudObjectKinds {
   static final String KIND_GLOBAL_WINDOW = "kind:global_window";
   static final String KIND_INTERVAL_WINDOW = "kind:interval_window";
+  static final String KIND_CUSTOM_WINDOW = "kind:custom_window";
```
Comment (robertwb) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectKinds.java:24]: Why is this needed?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+    @Override
+    public int hashCode() {
+      return Objects.hash(super.hashCode(), isBig);
+    }
+  }
+
+  private static class CustomWindowCoder extends CustomCoder<CustomWindow> {
+
+    private static final Coder<IntervalWindow> INTERVAL_WINDOW_CODER = IntervalWindow.getCoder();
+    private static final int REGISTER_BYTE_SIZE = 1234;
```
Comment (robertwb) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/TimestampPrefixingWindowCoderTest.java:70]: I'm not following what is meant by the `REGISTER_BYTE_SIZE` constant. `BYTE_SIZE_TO_REGISTER`?

File: model/fn-execution/src/main/resources/org/apache/beam/model/fnexecution/v1/standard_coders.yaml (hunk 13 above)
Comment (y1chi) [reply model/fn-execution/src/main/resources/org/apache/beam/model/fnexecution/v1/standard_coders.yaml:461]: yep, using interval_window coder here makes it easier to verify in standard coder test.

File: model/pipeline/src/main/proto/beam_runner_api.proto (hunk 15 above)
Comment (y1chi) [reply model/pipeline/src/main/proto/beam_runner_api.proto:945]: I think whether it should be length prefixed can be determined by runner with coder overwrite, since we are not limiting the wrapped window coder type, known coder can be also used(though it is unlikely in real use case, but it is handy for unit tests) which don't need length prefix.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectKinds.java (hunk 17 above)
Comment (y1chi) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectKinds.java:24]: For translating it as a known coder on runner side.

File: model/fn-execution/src/main/resources/org/apache/beam/model/fnexecution/v1/standard_coders.yaml (hunk 12 above)
Comment (y1chi) [reply model/fn-execution/src/main/resources/org/apache/beam/model/fnexecution/v1/standard_coders.yaml:460]: I think custom_window makes it a bit more clear that this is a coder for window type. timestamp_prefixed also works as long as sdk and runner agree with the urn I guess.

File: model/pipeline/src/main/proto/beam_runner_api.proto (hunk 14 above)
Comment (lukecwik) [reply model/pipeline/src/main/proto/beam_runner_api.proto:950]: I suggested copying it verbatim as that is the text in those segments already.

File: model/pipeline/src/main/proto/beam_runner_api.proto (hunk 15 above)
Comment (lukecwik) [reply model/pipeline/src/main/proto/beam_runner_api.proto:945]: I would rather have the length prefix to be a component as Yichi describes since this will allow other length prefixing schemes then the current one we have.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectKinds.java (hunk 17 above)
Comment (lukecwik) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectKinds.java:24]: It isn't needed since we shouldn't need to pass these in anymore to Dataflow since the beam proto -> DFE conversion should happen internally but I didn't ask to remove it since it was already done.
